src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -262,6 +262,56 @@ FabricInfo * RetrieveCurrentFabric(CommandHandler * aCommandHandler)
     return Server::GetInstance().GetFabricTable().FindFabricWithIndex(index);
 }
 
+void FailSafeCleanup(FailSafeContext & failSafeContext)
+{
+    FabricInfo * fabricInfo = Server::GetInstance().GetFabricTable().FindFabricWithIndex(failSafeContext.GetFabricIndex());
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: Call to FailSafeCleanup");
+
+    if (fabricInfo)
+    {
+        // If an AddNOC or UpdateNOC command has been successfully invoked, terminate all CASE sessions associated with the Fabric
+        // whose Fabric Index is recorded in the Fail-Safe context (see ArmFailSafe Command) by clearing any associated Secure
+        // Session Context at the Server.
+        if (failSafeContext.NocCommandHasBeenInvoked())
+        {
+            CASESessionManager * caseSessionManager = Server::GetInstance().GetCASESessionManager();
+
+            if (caseSessionManager)
+            {
+                caseSessionManager->ReleaseSession(fabricInfo->GetPeerId());
+            }
+        }
+
+        // If an AddNOC command had been successfully invoked, achieve the equivalent effect of invoking the RemoveFabric command
+        // against the Fabric Index stored in the Fail-Safe Context for the Fabric Index that was the subject of the AddNOC
+        // command.
+        if (failSafeContext.AddNocCommandHasBeenInvoked())
+        {
+            Server::GetInstance().GetFabricTable().Delete(fabricInfo->GetFabricIndex());
+        }
+
+        // If an UpdateNOC command had been successfully invoked, revert the state of operational key pair, NOC and ICAC for that
+        // Fabric to the state prior to the Fail-Safe timer being armed, for the Fabric Index that was the subject of the UpdateNOC
+        // command.
```

Comment:
Another piece that needs to be cleaned-up, is network config. Is there an issue to pre-save network config as a backup, or alternatively, only commit the new configs on CommissioningComplete? Please circle-back with Song Guo about support for this in Network Commissioning cluster  drive.

Reply:
That will be in separate PR, this PR is only about cleanup in OperationalCredentials cluster. Network-commissioning has non fabric scoped cleanup, need more clarification